### DISCUSSION TITLE
Add type/class invariants contract to Cpp2

### DISCUSSION THREAD

**Discussion by DyXel** [hsutter/cppfront#1021] [Suggestions]
Contracts already exist in Cpp2, and so do pre-conditions and post-conditions, but I would really like to be able to write my type contracts once and see the benefit on all member functions, which is exactly what invariants provide.

[p2900r4](http://wg21.link/p2900r4) says "... the following features are intentionally not included in this proposal; however, we expect these features to be proposed post-MVP extensions at a later time ...", followed by "... The ability to express class invariants. ...", so I think its in the spirit of cppfront to start experimenting on this early.

Personally I think we can start experimenting with a new contextual keyword (`invariant` as mentioned in the paper), and see where that leads us. Thoughts?

**Comment by AbhinavK00** [comment]
I think Herb has a metafunction in mind for just this kind of use case.
